### PR TITLE
Add an option for fast concatenation along a particular axis

### LINKED ISSUE #406: concatenation along specific axes (speed up combining of datasets)
Currently combining two data sets performs a check along all axes (blts,freqs,pols etc)  to allow arbitrary concatenation, but many times we just simply want to concatenate otherwise similar files along time or polarization.  This feature would allow concatenation of multiple data sets along specified axes ie
`UV1.concat(UV2,axis=pol)`
This will significantly speed up construction of datasets from multiple files.

### PR DESCRIPTION
**Pull request by bhazelton**
The code is fairly ready, but we need to add material to the tutorial and changelog about this.

## Description

Add an option to do fast concatenation along a particular axis (without much metadata checking)

## Motivation and Context

This is needed to speed up concatenation when it is known that the metadata matches exactly along the other axes.
closes #406

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the [contribution guide](.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] My change introduces new functionality that should be highlighted in the tutorial.
  - [x] I have updated the tutorial accordingly.
- [ ] If this is a bug fix, it includes a new test that breaks as a result of the bug (if possible)
- [ ] If this is a breaking change, it includes backwards compatibility and deprecation warnings (if possible)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass in both python 2 and python 3.
- [x] My change requires an update to the [CHANGELOG](CHANGELOG.md).
  - [x] I have updated the [CHANGELOG](CHANGELOG.md) accordingly.

### CODE REVIEW

File: pyuvdata/uvdata.py [hunk 1]
```diff
@@ -1185,6 +1185,144 @@ def __iadd__(self, other):
...
+        # Check that both objects are UVData and valid
+        this.check(check_extra=check_extra, run_check_acceptability=run_check_acceptability)
+        if not issubclass(other.__class__, this.__class__):
+            if not issubclass(this.__class__, other.__class__):
+                raise ValueError('Only UVData (or subclass) objects can be '
+                                 'added to a UVData (or subclass) object')
+        other.check(check_extra=check_extra, run_check_acceptability=run_check_acceptability)
+
+        allowed_axes = ['blt', 'freq', 'polarization']
+        if axis is not None:
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: Looks like `axis` is required, so I'm not sure this `if` is necessary.

File: pyuvdata/uvdata.py [hunk 2]
```diff
@@ -1751,7 +1894,11 @@ def read_uvfits(self, filename, antenna_nums=None, antenna_names=None,
                                     run_check=run_check, check_extra=check_extra,
                                     run_check_acceptability=run_check_acceptability,
                                     keep_all_metadata=keep_all_metadata)
-                    self += uv2
+                    if axis is not None:
+                        self.fast_concat(uv2, axis, run_check=True, check_extra=True,
+                                         run_check_acceptability=True, inplace=True)
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: Should the check keywords be based on what was passed into this function?

File: pyuvdata/uvdata.py [hunk 3]
```diff
@@ -1843,7 +1996,11 @@ def read_ms(self, filepath, run_check=True, check_extra=True,
                     uv2.read_ms(f, run_check=run_check, check_extra=check_extra,
                                 run_check_acceptability=run_check_acceptability,
                                 data_column=data_column, pol_order=pol_order)
-                    self += uv2
+                    if axis is not None:
+                        self.fast_concat(uv2, axis, run_check=True, check_extra=True,
+                                         run_check_acceptability=True, inplace=True)
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: Same as above

File: pyuvdata/uvdata.py [hunk 4]
```diff
@@ -1882,7 +2045,11 @@ def read_fhd(self, filelist, use_model=False, run_check=True, check_extra=True,
                     uv2.read_fhd(f, use_model=use_model, run_check=run_check,
                                  check_extra=check_extra,
                                  run_check_acceptability=run_check_acceptability)
-                    self += uv2
+                    if axis is not None:
+                        self.fast_concat(uv2, axis, run_check=True, check_extra=True,
+                                         run_check_acceptability=True, inplace=True)
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: Same as above

File: pyuvdata/uvdata.py [hunk 5]
```diff
@@ -1935,7 +2108,7 @@ def read_miriad(self, filepath, antenna_nums=None, ant_str=None, bls=None,
         from . import miriad
         if isinstance(filepath, (list, tuple)):
             if not read_data:
-                raise ValueError('read_data cannot be False for a list of uvfits files')
+                raise ValueError('read_data cannot be False for a list of miriad files')
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: whoops, good catch.

File: pyuvdata/uvdata.py [hunk 6]
```diff
@@ -1952,7 +2125,11 @@ def read_miriad(self, filepath, antenna_nums=None, ant_str=None, bls=None,
                                     phase_type=phase_type, antenna_nums=antenna_nums,
                                     ant_str=ant_str, bls=bls,
                                     polarizations=polarizations, time_range=time_range)
-                    self += uv2
+                    if axis is not None:
+                        self.fast_concat(uv2, axis, run_check=True, check_extra=True,
+                                         run_check_acceptability=True, inplace=True)
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: same as above

File: pyuvdata/uvdata.py [hunk 7]
```diff
@@ -2090,7 +2273,11 @@ def read_uvh5(self, filename, antenna_nums=None, antenna_names=None,
                                   run_check_acceptability=run_check_acceptability,
                                   data_array_dtype=data_array_dtype,
                                   keep_all_metadata=keep_all_metadata)
-                    self += uv2
+                    if axis is not None:
+                        self.fast_concat(uv2, axis, run_check=True, check_extra=True,
+                                         run_check_acceptability=True, inplace=True)
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: same as above

File: pyuvdata/uvdata.py [hunk 8]
```diff
@@ -2069,7 +2252,7 @@ def read_uvh5(self, filename, antenna_nums=None, antenna_names=None,
             if not read_data and len(filename) > 1:
                 raise ValueError('read_data cannot be False for a list of uvh5 files')
 
-            self.read_uvh5(filename[0], antenna_nums=antenna_nums,
+            self.read_uvh5(filename[0], axis=axis, antenna_nums=antenna_nums,
```
Comment (adampbeardsley) [inline pyuvdata/uvdata.py]: Why is axis getting passed in here (and on line 2267)? I don't think it changes anything, but it's inconsistent with the other read functions.

**Review by adampbeardsley — COMMENTED** [review]
Looks good so far. I added a few comments to clean things up a little. Then it needs the tutorial and documentation updates.

File: pyuvdata/uvdata.py (hunk 1 above)
Comment (bhazelton) [reply pyuvdata/uvdata.py]: yep, thanks!

File: pyuvdata/uvdata.py (hunk 2 above)
Comment (bhazelton) [reply pyuvdata/uvdata.py]: yes, thank you!

File: pyuvdata/uvdata.py (hunk 8 above)
Comment (bhazelton) [reply pyuvdata/uvdata.py]: you're right, I'm not sure why it's there, I'll take it out.

**Comment by plaplant** [comment]
For posterity, attached is a plot showing a comparison in timing between the `__add__` and `fast_concat` for reading in a list of files.  These are H2C data files in uvh5 format. Both scale more steeply than linearly, but clearly `fast_concat` is winning by a mile.

times

**Comment by bhazelton** [comment]
@plaplant Thank you! That's a great plot.

**Review by adampbeardsley — APPROVED** [review]
Looks great to me! Thanks @bhazelton and @plaplant for getting this done!

**Pull request merged**